### PR TITLE
Allow to configure and report key modifiers in `TapTool` and pointer events

### LINKED ISSUE #12532: [FEATURE] Access to know whether user ctrl-clicks, shift-clicks, or command-clicks in CustomJS callback
### Problem description

I am trying to make a tap tool with two different behaviors depending on whether the user is clicking on a point or control + clicking a point.

I would like this so that the user can tap for one type of behavior, but also control + click to open a URL associated with the point.

My example to start with that does not have this feature, it just opens the link when you press the points:

```python
from bokeh.models import CustomJS, TapTool, ColumnDataSource
from bokeh.plotting import figure, show

cds = ColumnDataSource({
    'x': [1, 2, 3],
    'y': ['a', 'b', 'c'],
    'link': ['www.google.com', 'www.bing.com', 'www.cnn.com']
})
plot = figure(width=300, height=300, y_range=['a', 'b', 'c'])
plot.scatter(x='x', y='y', color="#fa9fb5", source=cds,
             marker='triangle_pin', size=15)

code = """
window.open('http://'.concat(cb_data.source.data.link[cb_data.source.inspected.indices]), '_blank')
"""
callback = CustomJS(args=dict(), code=code)
tap_tool = TapTool(behavior='inspect', callback=callback)
plot.add_tools(tap_tool)

show(plot)
```

I want to adjust the callback code to be something like this:

```python
code = """
console.log(window)
console.log(MouseEvent.ctrlKey)
if (MouseEvent.ctrlKey) {
    window.open('http://'.concat(cb_data.source.data.link[cb_data.source.inspected.indices]), '_blank')
} else {
   // do something different
}
"""
```

### Feature description

Create a variable in CustomJS that can communicate the type of click.

### Potential alternatives

One potential alternative is a custom extension.

I came up with a workaround that does not use a custom extension, which I am sure is not ideal:

```python
from bokeh.models import CustomJS, TapTool, ColumnDataSource, Tabs, Panel, Div
from bokeh.plotting import figure, show

cds = ColumnDataSource({
    'x': [1, 2, 3],
    'y': ['a', 'b', 'c'],
    'link': ['www.google.com', 'www.bing.com', 'www.cnn.com']
})
plot = figure(width=300, height=300, y_range=['a', 'b', 'c'])
plot.scatter(x='x', y='y', color="#fa9fb5", source=cds, marker='triangle_pin', size=15)

code_1 = """
window.onclick = function () {
    window.ctrl_press = event.ctrlKey || event.metaKey
}
"""
callback_1 = CustomJS(code=code_1)

code_2 = """
if (window.ctrl_press) {
    window.open('http://'.concat(cb_data.source.data.link[cb_data.source.inspected.indices]), '_blank')
}
"""
callback_2 = CustomJS(code=code_2)

tap_tool = TapTool(behavior='inspect', callback=callback_2)
plot.add_tools(tap_tool)

panels = [Panel(child=Div(text='Welcome!'), title='Intro'), Panel(child=plot, title='Example')]
tabs = Tabs(tabs=panels, sizing_mode='stretch_width')
tabs.js_on_change('active', callback_1)

show(tabs)
```

When the user changes tabs, a callback is set to make a property stored in `window` equal to whether the control or command key was pressed in the window when a click happens. Now, the TapTool callback has access to this property in `window`.

### Additional information
I am using Bokeh 2.4.2 in the examples.

This is based on the bokeh discourse discussion [How to see if user shift clicked or control clicked or command clicked on Tap Tool javascript callback](https://discourse.bokeh.org/t/how-to-see-if-user-shift-clicked-or-control-clicked-or-command-clicked-on-tap-tool-javascript-callback/9679)

Thanks in advance!

### PR DESCRIPTION
**Pull request by mattpap**
This adds two things:

- Allows to configure `TapTool` with key modifiers required to trigger the tool. This is specifically useful in `inspect` mode, less in `select` mode, which has intrinsic meaning for key modifiers, derived from other select tools.
- Allows to report key modifiers in `TapTool`'s callback data and allows the same for pointer events `Tap`, etc.

fixes #12532

### CODE REVIEW

File: src/bokeh/models/tools.py [hunk 1]
```diff
@@ -676,6 +679,16 @@ def __init__(self, *args, **kwargs) -> None:
     either a single or double tap.
     """)
 
+    modifiers = Struct(shift=Optional(Bool), ctrl=Optional(Bool), alt=Optional(Bool), default={}, help="""
+    Allows to configure a combination of modifier keys, which need to
+    be pressed during the selected gesture for this tool to trigger.
+
+    .. note::
```
Comment (bryevdv) [inline src/bokeh/models/tools.py]: Maybe

```suggestion
    .. warn::
```

File: examples/interaction/tools/tap_inspect.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import numpy as np
```
Comment (bryevdv) [inline examples/interaction/tools/tap_inspect.py]: Maybe `tap_inspect_keymod.py` just to highlight that this example covers key modifiers?

**Review by bryevdv — APPROVED** [review]
LGTM modulo two small suggestions

**Comment by nlosacco** [comment]
but this still only allows for a single action per each TapTool, right?
I mean, it's not possible to set up different actions for different modifiers with a single TapTool, is it?

**Pull request merged**